src/wmtk/invariants/Swap2dUnroundedVertexInvariant.cpp

```diff
@@ -0,0 +1,42 @@
+#include "Swap2dUnroundedVertexInvariant.hpp"
+#include <wmtk/Mesh.hpp>
+#include <wmtk/function/utils/amips.hpp>
+#include <wmtk/utils/orient.hpp>
+
+namespace wmtk {
+Swap2dUnroundedVertexInvariant::Swap2dUnroundedVertexInvariant(
+    const Mesh& m,
+    const attribute::TypedAttributeHandle<Rational>& coordinate)
+    : Invariant(m, true, false, false)
+    , m_coordinate_handle(coordinate)
+{}
+
+bool Swap2dUnroundedVertexInvariant::before(const simplex::Simplex& t) const
+{
+    constexpr static PrimitiveType PV = PrimitiveType::Vertex;
+    constexpr static PrimitiveType PE = PrimitiveType::Edge;
+    constexpr static PrimitiveType PF = PrimitiveType::Triangle;
+
+    auto accessor = mesh().create_const_accessor(m_coordinate_handle);
+    int dim = mesh().get_attribute_dimension(m_coordinate_handle);
+
+    // get the coords of the vertices
+    // input face end points
+    const std::array<Tuple, 4> vs = {
```

Comment:
You can use `simplex::faces_single_dimension` for that. Or even better, use `Mesh::orient_vertices()` which is more efficient.

Reply:
I need a vertex in the opposite triangle so the last switch is needed anyway. and this code will be replaced by a swap2d before check, I will change it there